include/zephyr/storage/storage_area/storage_area.h

```diff
@@ -0,0 +1,216 @@
+ *
+ *
+ * There following methods area exposed:
+ *   storage_area_read(),	** read data **
+ *   storage_area_readv(),	** read data vector **
+ *   storage_area_write(),	** write data **
+ *   storage_area_writev(),	** write data vector **
+ *   storage_area_erase(),	** erase (in erase block addressing) **
+ *   storage_area_ioctl()	** used for e.g. getting xip addresses **
+ *
+ * The subsystem is easy extendable to create custom (virtual) storage areas
+ * that consist of e.g. a combination of flash and ram, a encrypted storage
+ * area, ...
+ *
+ * A storage area is defined e.g. for flash:
+ * struct storage_area_flash fa =
+ *	flash_storage_area(dev, start, xip_address, write_size, erase_size,
+ *			   size, properties);
+ * struct storage_area *area = &fa.area;
+ *
+ * For other storage devices (eeprom, ram, disk, ...) similar macros are
+ * defines, but they might differ slightly.
+ *
+ * The write_size, erase_size, ... are declarations of how the storage_area
+ * will be used The write_size is limited to a power of 2, erase_size should
+ * be a multiple of write_size and size should be a multiple of erase_size.
+ * The macro definitions xxxxx_storage_area(...) checks these conditions but
+ * always succeeds. Trying to use a badly sized storage area will result in
+ * failure of any of the exposed methods.
+ *
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_H_
+#define ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_H_
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Storage_area interface
+ * @defgroup storage_area_interface Storage_area interface
+ * @ingroup storage_apis
+ * @{
+ */
+
+struct storage_area;
+
+struct storage_area_iovec { /* storage area io vector */
+	void *data;         /* pointer to data */
+	size_t len;         /* data length */
+};
+
+enum storage_area_properties_mask {
+	SA_PROP_READONLY = 0x0001,
+	SA_PROP_FOVRWRITE = 0x0002, /* full overwrite (ram, rram, ...) */
+	SA_PROP_LOVRWRITE = 0x0004, /* limited overwrite (nor flash) */
+	SA_PROP_ZEROERASE = 0x0008, /* erased value is 0x00 */
+	SA_PROP_AUTOERASE = 0x0010, /* erase while writing */
+};
+
+enum storage_area_ioctl_cmd {
+	SA_IOCTL_NONE,
+	SA_IOCTL_XIPADDRESS, /* retrieve the storage area xip address */
+};
+
+/**
+ * @brief storage_area API
+ *
+ * API to access storage area.
+ */
+struct storage_area_api {
+	int (*readv)(const struct storage_area *area, size_t start,
+		     const struct storage_area_iovec *iovec, size_t iovcnt);
+	int (*writev)(const struct storage_area *area, size_t start,
+		    const struct storage_area_iovec *iovec, size_t iovcnt);
+	int (*erase)(const struct storage_area *area, size_t start, size_t bcnt);
+	int (*ioctl)(const struct storage_area *area,
+		     enum storage_area_ioctl_cmd cmd, void *data);
+};
+
+/**
+ * @brief storage_area struct
+ */
+struct storage_area {
+	const struct storage_area_api *api;
+	size_t write_size;
+	size_t erase_size;
+	size_t erase_blocks;
```

Comment:
There should be rather overall device size here, which is even indicated by initialization of disk_storage_area, flash_storage_area, and eeprom_storage area  which all have to populate this by division of size by erase_size.
Additionally, in code of disk and eeprom you have to put the size back by multiplying them back, and there is no invocation of STORAGE_AREA_ERASESIZE anywhere and direct usage of erase_blocks is only in  validations and in flash erase, which means that there is more cases for direct use of size rather than erase_blocks, which can be calculated from true, or emulated, erase size and size.

Note also that https://github.com/zephyrproject-rtos/zephyr/issues/71270 is in path of making erase optional for drivers, i.e. returning -ENOTSUP, some drivers will no longer return pages layout nor hold information on device erase page. That benefits devices like MRAM in Ambiq or RRAM in Nordic, and many other devices where code not needed for these drivers to work and be removed and benefit users by reduction of code flash usage taken by vendor code to support these devices.

Reply:
Whether this should be `size` or `erase_blocks` should be determined based on use. A typical bootloader will want to know the number or erase-blocks to find out how much space is required for the update status and to walk over while copying/moving. By using the number of erase-blocks the application code might be increased a little, but at the advantage of reduced bootloader size.